services/appflowy-collaborate/src/group/group_init.rs

```diff
@@ -285,6 +300,89 @@ impl EditState {
   }
 }
 
+struct CollabUpdateStreamingImpl {
+  workspace_id: String,
+  object_id: String,
+  collab_redis_stream: Arc<CollabRedisStream>,
+  mem_cache: Mutex<Vec<Vec<u8>>>,
+  update_stream: Mutex<Option<StreamGroup>>,
+}
+
+impl CollabUpdateStreamingImpl {
+  fn new(workspace_id: &str, object_id: &str, collab_redis_stream: Arc<CollabRedisStream>) -> Self {
+    Self {
+      workspace_id: workspace_id.to_string(),
+      object_id: object_id.to_string(),
+      collab_redis_stream,
+      mem_cache: Default::default(),
+      update_stream: Default::default(),
+    }
+  }
+
+  /// Initialize redis update stream
+  /// After initialize the redis update stream, all pending updates will be sent immediately.
+  async fn init_update_stream(&self) -> Result<(), StreamError> {
+    let mut redis_update_stream = self
+      .collab_redis_stream
+      .collab_update_stream(
+        &self.workspace_id,
+        &self.object_id,
+        "collaborate_update_producer",
+      )
+      .await?;
+
+    // Send all pending updates
+    send_pending_updates(&mut redis_update_stream, &self.mem_cache).await;
+    *self.update_stream.lock().await = Some(redis_update_stream);
+    Ok(())
+  }
+}
+
+#[async_trait]
+impl CollabUpdateStreaming for CollabUpdateStreamingImpl {
+  async fn send_update(&self, update: Vec<u8>) {
+    if let Some(redis_update_stream) = self.update_stream.lock().await.deref_mut() {
+      if let Err(err) = redis_update_stream
+        .insert_message(CollabUpdateEvent::UpdateV1 {
+          encode_update: update,
+        })
+        .await
+      {
+        //FIXME: If fail to insert update to redis stream, which will cause missing update issue.
+        error!(
+          "{} fail to insert update into redis stream. error: {}",
+          self.object_id, err
+        );
+      }
+
+      return;
+    }
+
+    // Cache update if the redis update stream is not ready.
+    self.mem_cache.lock().await.push(update);
+    self.init_update_stream().await;
+  }
+}
+
+async fn send_pending_updates(redis_update_stream: &mut StreamGroup, cache: &Mutex<Vec<Vec<u8>>>) {
+  let messages = cache
+    .lock()
+    .await
+    .drain(..)
+    .into_iter()
+    .flat_map(|update| {
+      StreamBinary::try_from(CollabUpdateEvent::UpdateV1 {
+        encode_update: update,
+      })
+      .ok()
+    })
+    .collect::<Vec<StreamBinary>>();
+  // FIXME: How to handle the situation when there are too many messages that need to be sent
+  if let Err(err) = redis_update_stream.insert_messages(messages).await {
```

Comment:
Yrs has method called [merge_updates](https://docs.rs/yrs/latest/yrs/fn.merge_updates_v1.html), so you could squash multiple updates into one).